Add add_subgraph and remove_subgraph functionality to REAPI

milroy opened the pull request:
This PR exposes `grow` functions in the C and C++ REAPI bindings.

The `grow` functionality passes a JGF subgraph including the path from the cluster vertex to the subgraph root. For example, a JGF subgraph with a new node (`newnode`) and subnode resources to `cluster0` at `rack0` includes the cluster and rack vertices as well as the induced edges. The disadvantage of this approach is that the vertex metadata in the JGF subgraph needs to be sufficiently specified to identify the vertices that already exist in the graph (e.g., `cluster0`).

milroy committed:
reader: update JGF reader comment

Problem: the code comment on the unpack_at function indicates that there
is no copy constructor for planner. That hasn't been true for years.

Update the comment.

milroy committed:
reader: update JGF reader comment

Problem: the code comment on the unpack_at function indicates that there
is no copy constructor for planner. That hasn't been true for years.

Update the comment.

milroy committed:
reapi: add C++ API add_subgraph functionality

Problem: there is no C++ API function or functionality to add a subgraph
to an existing resource graph.

Add the C++ API function and implementation to call the JGF reader and
reinitialize the traverser to add pruning filters and update pruning
filter counts.

milroy committed:
reapi: add C++ API add_subgraph functionality

Problem: there is no C++ API function or functionality to add a subgraph
to an existing resource graph.

Add the C++ API function and implementation to call the JGF reader and
reinitialize the traverser to add pruning filters and update pruning
filter counts.

milroy committed:
resource-query: add_subgraph, remove_subgraph API calls

Problem: resource-query uses its own methods to add and remove
subgraphs.

Update the methods to use the C++ API calls.

milroy committed:
resource-query: add_subgraph, remove_subgraph API calls

Problem: resource-query uses its own methods to add and remove
subgraphs.

Update the methods to use the C++ API calls.

milroy committed:
reapi: fix C API error message output

Problem: getting the CLI error message does not check for non-null ctx.

Add the check. Also return an owned string, and surface the accumulated
error when rqt is null."

milroy committed:
reapi: fix C API error message output

Problem: getting the CLI error message does not check for non-null ctx.

Add the check.

milroy committed:
reapi: add add_subgraph function to C API

Problem: there is no C API function to add a subgraph to an existing
resource graph.

Add the C API function.

milroy committed:
reapi: add add_subgraph function to C API

Problem: there is no C API function to add a subgraph to an existing
resource graph.

Add the C API function.

milroy committed:
reapi: add remove_subgraph function to C API

Problem: there is no C API function to remove a subgraph from an existing
resource graph.

Add the C API function.

milroy committed:
reapi: add remove_subgraph function to C API

Problem: there is no C API function to remove a subgraph from an existing
resource graph.

Add the C API function.

milroy committed:
reapi: add C++ API remove_subgraph functionality

Problem: there is no C++ API function or functionality to remove a
subgraph from an existing resource graph.

Add the C++ API function and implementation to remove a subgraph rooted
at the given path to a vertex and reinitialize the traverser to update
pruning filter counts.

milroy committed:
reapi: add C++ API remove_subgraph functionality

Problem: there is no C++ API function or functionality to remove a
subgraph from an existing resource graph.

Add the C++ API function and implementation to remove a subgraph rooted
at the given path to a vertex and reinitialize the traverser to update
pruning filter counts.

vsoch commented:
This is tested (and the basics are working in fluxion-go)! :partying_face: 

We start with cluster [tiny](https://github.com/flux-framework/fluxion-go/blob/add-grow-support/cmd/test/data/tiny.json), one rack and two nodes (node0 and node1). We demonstrate that if we ask for 4 nodes, we cannot be satisfied:

```console
Asking to MatchSatisfy 4 nodes (not possible)

        ----Match Satisfy output---
satisfied: false
error: <nil>
```

We then ask fluxion to grow from 2 to 4 nodes. We do that with a [new nodes request](https://github.com/flux-framework/fluxion-go/blob/add-grow-support/cmd/test/data/grow/new-nodes.json) that includes an existing path in the graph, `tiny0->rack0` but then defines two new nodes, `node2` and `node3`.

```console
🍔 Asking to Grow from 2 to 4 Nodes
Grow request return value: <nil>
```

We can get some verification that the graph node has node0-3 (4 nodes) by asking for 4 nodes with satisfy again. We are satisfied!

```console
Asking to MatchSatisfy 4 nodes (now IS possible)

        ----Match Satisfy output---
satisfied: true
error: <nil>
```

We now want to test shrink. Shrink takes the node path that we should prune at. For this case, we just prune off one node. 
```console
🥕 Asking to Shrink from 4 to 3 Nodes
Shrink request return value: <nil>
```
When we have 3 nodes we ask for 4 again, and we are no longer satisfied.

```console
Asking to MatchSatisfy 4 nodes (again, not possible)

        ----Match Satisfy output---
satisfied: false
error: <nil>
```

The testing shown above is [here in GitHub CI](https://github.com/flux-framework/fluxion-go/actions/runs/12780828020/job/35627749377) (shown across OSes) and that full PR is [here](https://github.com/flux-framework/fluxion-go/pull/12) and will just need to be updated when the branch here is merged (or we can work from this branch if that isn't going to happen soon - I can build/deploy a custom container that has it. 

## How will shrink work in Kubernetes?

High level - I'm thinking through the shrink design for a cluster in Kubernetes, and I think we have two use cases (that warrant different design strategies):

### 1. A need to shrink down in increments of 1

This will work fine to prune single nodes.

### 2. A need to shrink down in unknown increments (>1)

This could be a lot of requests to fluxion, for example, if we want to shrink down by 10, 20, or more nodes at once. We have a few options, I think. I'll think through each one.

- We can expose a function in fluxion-go that takes a list of nodes, and then (within the same call to fluxion-go) makes multiple calls to fluxion. That mostly replaces the need to do many grpc requests (one per node) with just one (to handle many requests, internally). 

- If we have an understanding of the increments, we can design a cluster graph with abstract levels of racks, where each rack is a group of nodes that are intended to be brought down together. Maybe we would design that graph based on topology. That way, we can do one request to trim the rack, and it will cut off all the nodes.

I haven't looked at what fluxion is doing in terms of the actual shrink (happening during a traversal?) but if there is an operation that can handle multiple cuts at once (e.g., done during one traversal) that could be an idea. But based on my impression of infrequent scaling, I don't really think optimizing this up the wazoo right now is that necessary.

## How will grow design work in Kubernetes?

For the first case (a flat topology that has nodes added to it) we likely just need to get the highest node identifier in the graph, and then we generate the json request that adds 1 to that (and however many we need). That can be stored as a variable somewhere, and on restart it can be calculated again from the live cluster.

For the second case that has multiple racks each with children, we would likely apply the same strategy as above, but on the level of the rack, and then the number of children under a rack is constant. We would calculate the node indices based on the number of racks and expected nodes per each. In the case that we have different numbers of nodes per rack (e.g., imagine different applications requiring different sized increments) we could label the rack with metadata that says exactly the number of children that are there.

Anyway - there are many ways to skin an avocado! Thanks for finishing this up @milroy I'm super pumped to see it working, and (TBA) to get it merged and deployed and into some of our projects!

vsoch commented:
@trws when you are back in business :briefcase: could you review this PR? And no rush if you are still on travel or otherwise - I can continue building from the PR branch.

milroy commented:
> @trws when you are back in business 💼 could you review this PR? And no rush if you are still on travel or otherwise - I can continue building from the PR branch.

We'll need to discuss what to do about testing. These API changes look like dead code to coverage.

vsoch commented on resource/reapi/bindings/c++/reapi_cli_impl.hpp:
```diff
@@ -722,6 +748,73 @@ int resource_query_t::remove_job (const uint64_t jobid, const std::string &R, bo
+        m_err_msg += ": ERROR: can't create JGF reader\n";
+        return rc;
+    }
+    if ((rc = reader->remove_subgraph (db->resource_graph, db->metadata, subgraph_path)) != 0) {
```
Is `remove_subgraph` on the reader or the traverser? I am looking [here](https://github.com/flux-framework/flux-sched/blob/e1af864ee48d503314fd6acc075102151cbdc0ee/resource/traversers/dfu_impl.hpp#L611).

milroy replied:
I moved `remove_subgraph` from the reader to the traverser in a PR that was merged before the rebase here. Good catch!

milroy committed:
reapi unit tests: add catch2 unit testing

Problem: the REAPI C API doesn't have unit testing.

Add a Catch2-based unit test for future extension.

milroy committed:
resource module: remove_subgraph functionality

Problem: there is no functionality in the resource match module to remove a
subgraph from an existing resource graph.

Add the callback and implementation to remove a subgraph rooted
at the given path to a vertex and reinitialize the traverser to update
pruning filter counts.

milroy committed:
resource module: remove_subgraph functionality

Problem: there is no functionality in the resource match module to remove a
subgraph from an existing resource graph.

Add the callback and implementation to remove a subgraph rooted
at the given path to a vertex and reinitialize the traverser to update
pruning filter counts.

milroy committed:
testsuite: update expected error output

Problem: the current expected output error message differs from the new
format.

Update the expected output to the updated format.

milroy committed:
testsuite: update expected error output

Problem: the current expected output error message differs from the new
format.

Update the expected output to the updated format.

milroy committed:
resource module: add add_subgraph functionality

Problem: there is no existing functionality to add a subgraph to an
existing resource graph in the resource module.

Add implementation to add a subgraph and reinitialize the traverser to
add pruning filters and update counts.

milroy committed:
resource module: add add_subgraph functionality

Problem: there is no existing functionality to add a subgraph to an
existing resource graph in the resource module.

Add implementation to add a subgraph and reinitialize the traverser to
add pruning filters and update counts.

milroy committed:
ion-resource: add_subgraph functionality

Problem: flux ion-resource lacks the functionality to add subgraphs via
RPC to the resource module.

Add the required subcommand, parsing, and RPC.

milroy committed:
ion-resource: add_subgraph functionality

Problem: flux ion-resource lacks the functionality to add subgraphs via
RPC to the resource module.

Add the required subcommand, parsing, and RPC.

milroy committed:
ion-resource: remove_subgraph functionality

Problem: flux ion-resource lacks the functionality to remove subgraphs via
RPC to the resource module.

Add the required subcommand, parsing, and RPC.

milroy committed:
ion-resource: remove_subgraph functionality

Problem: flux ion-resource lacks the functionality to remove subgraphs via
RPC to the resource module.

Add the required subcommand, parsing, and RPC.

milroy committed:
testsuite: add tests for subgraph add/remove

Problem: the sharness testsuite lacks the capability to test the API
functions added in this PR.

Add tests for successful addition, removal, and error checking.

milroy committed:
testsuite: add tests for subgraph add/remove

Problem: the sharness testsuite lacks the capability to test the API
functions added in this PR.

Add tests for successful addition, removal, and error checking.

trws commented on resource/modules/resource.cpp:
```diff
@@ -1030,6 +1045,62 @@ static void get_property_request_cb (flux_t *h,
+        goto error;
+
+    R_s = R_subgraph;
+    if (run_add_subgraph (ctx, R_s) != 0) {
```
```suggestion
    std::string errmsg = "";
    char *R_subgraph = NULL;

    if (flux_request_unpack (msg, NULL, "{s:s}", "subgraph", &R_subgraph) < 0)
        goto error;

    if (run_add_subgraph (ctx, R_subgraph) != 0) {
```

milroy replied:
Implemented.

trws commented on resource/modules/resource.cpp:
```diff
@@ -1030,6 +1045,62 @@ static void get_property_request_cb (flux_t *h,
+        goto error;
+
+    s_path = subgraph_path;
+    if (run_remove_subgraph (ctx, s_path) != 0) {
```
```suggestion
    std::string errmsg = "";
    char *subgraph_path = NULL;

    if (flux_request_unpack (msg, NULL, "{s:s}", "subgraph_path", &subgraph_path) < 0)
        goto error;

    if (run_remove_subgraph (ctx, subgraph_path != 0) {
```

milroy replied:
Implemented.

trws commented on resource/modules/resource_match.cpp:
```diff
@@ -1685,6 +1685,67 @@ int run_find (std::shared_ptr<resource_ctx_t> &ctx,
+    }
+    // Could be adding new resource types, so open the interner storage before adding the subgraph
+    subsystem_t::storage_t::open ();
+    resource_type_t::storage_t::open ();
```
This is the right logic, but should be using the RAII wrapper I added to the interner.  Not sure it existed when you wrote this, but it would increase our exception safety here.  Should be just change these for:


```suggestion
    auto subsystem_open = subsystem_t::storage_t::open_for_scope ();
    auto resource_open = resource_t::storage_t::open_for_scope ();
```

and then remove the finalizes below.

trws commented on resource/reapi/bindings/c++/reapi_cli_impl.hpp:
```diff
@@ -739,6 +769,59 @@ int resource_query_t::remove_job (const uint64_t jobid, const std::string &R, bo
+    }
+    // Could be adding new resource types, so open the interner storage before adding the subgraph
+    subsystem_t::storage_t::open ();
+    resource_type_t::storage_t::open ();
```
Same as the module implementation.

jameshcorbett commented on resource/reapi/test/reapi_cli_test.cpp:
```diff
@@ -0,0 +1,137 @@
+#include <fstream>
+#include <iostream>
+
+namespace Flux::resource_model::detail {
```
this test needs to be added to the build system?

jameshcorbett replied:
It also fails for me when I ran it locally.

milroy replied:
I totally reworked the tests. I fixed this as part of the refactor.

jameshcorbett commented on resource/modules/resource_match.cpp:
```diff
@@ -1745,6 +1745,68 @@ int run_find (std::shared_ptr<resource_ctx_t> &ctx,
+    {
+        auto subsystem_open = subsystem_t::storage_t::open_for_scope ();
+        auto resource_open = resource_type_t::storage_t::open_for_scope ();
+    }
```
Since these are scoped though, won't they be destructed and closed immediately?

trws replied:
Yes, they need to be at function scope rather than a sub-scope like that.

milroy replied:
Yes, this was an error that I realized shortly after pushing. I fixed it but didn't push the changes.

jameshcorbett commented on resource/reapi/test/reapi_cli_test.cpp:
```diff
@@ -0,0 +1,137 @@
+    std::ifstream resource_file ("../../../t/data/resource/grugs/tiny.graphml");
+    std::ifstream jobspec_file ("../../../t/data/resource/jobspecs/basics/test001.yaml");
+    std::ifstream jobspec_file2 ("../../../t/data/resource/jobspecs/basics/test011.yaml");
+    match_op_t match_op = match_op_t::MATCH_ALLOCATE;
```
Use `SHARNESS_TEST_SRCDIR` here instead?

milroy replied:
I totally reworked the tests. I fixed this as part of the refactor.

jameshcorbett commented on resource/modules/resource_match.cpp:
```diff
@@ -1745,6 +1745,68 @@ int run_find (std::shared_ptr<resource_ctx_t> &ctx,
+    }
+    // Update total counts:
+    rc = ctx->traverser->initialize ();
+    flux_log (ctx->h, LOG_DEBUG, "successfully added subgraph %s", R_subgraph.c_str ());
```
delete this line? It's logged without even checking the value of `rc`.

trws replied:
Probably copy up the equivalent block from `run_remove_subgraph`: https://github.com/flux-framework/flux-sched/pull/1316/changes/BASE..408a2c01ce5270815c701964ba9821c6944320c4#diff-48628982be5211b25532174bc920cdedb26fbc5cee95e74f9f3288c1390b4f2eR1797-R1804

milroy replied:
Good catch! Fixed.

jameshcorbett commented on resource/reapi/bindings/c++/reapi.hpp:
```diff
@@ -194,6 +194,38 @@ class reapi_t {
+     */
+    static int add_subgraph (void *h, const std::string &R_subgraph)
+    {
+        return -1;
```
For not implemented methods @garlick set `errno = ENOSYS` elsewhere, I think maybe we should follow that pattern? E.g. commit [67952773d9a903f2a3582e19b19b1fe196fc9427](https://github.com/flux-framework/flux-sched/pull/1489/changes/67952773d9a903f2a3582e19b19b1fe196fc9427)

milroy replied:
The not-yet-implemented methods are set as you describe in the implementation header file (`reapi_cli_impl.hpp`). The ones with `return -1` are actually implemented.

trws replied:
these are somewhat strange.  Really they should be abstract methods, but they aren't methods.  They're in a strange limbo place because of the way reapi is set up.  In principle they should probably be just `throw std::logic_error("Function not implemented");` or something but since the rest of these are all like this, I'd say leave it as is right now.

milroy committed:
reapi: fix null dereference

Problem: reapi_cli_match() dereferences ctx->rqt before checking
it is non-null, so calling match on an uninitialized context segfaults.

Add the correct null checks.

Assisted-by: Claude:claude-opus-4.8

milroy committed:
reapi unit tests: cover subgraph add and remove

Problem: there's not Catch2 unit testing for the new functionality.

Add Catch2 coverage for the new add_subgraph/remove_subgraph API to the
C and C++ REAPI binding tests (reapi_cli_test_c.cpp and
reapi_cli_test_cxx.cpp): add a subgraph to a JGF graph and remove it,
and confirm both calls fail cleanly before initialization and on empty
input.

Add a C-binding regression test that adds a subgraph introducing a new
resource type. This covers the path where the interner storage must
stay open across the reader's unpack_at; additions that reuse existing
types do not exercise it, so the test guards against the storage being
closed too early.

Assisted-by: Claude:claude-opus-4.8

jameshcorbett commented on resource/modules/resource_match.cpp:
```diff
@@ -1745,6 +1745,46 @@ int run_find (std::shared_ptr<resource_ctx_t> &ctx,
+    if ((reader = create_resource_reader ("jgf")) == nullptr) {
+        flux_log (ctx->h,
+                  LOG_ERR,
+                  "creating JGF reader failed: %s",
+                  reader->err_message ().c_str ());
+        goto done;
```
`reader` is null here

milroy replied:
Good catch; fixed.

jameshcorbett reviewed (approved):
One thing, otherwise this LGTM

trws commented on resource/reapi/bindings/c/reapi_cli.cpp:
```diff
@@ -203,9 +203,15 @@ extern "C" int reapi_cli_match (reapi_cli_ctx_t *ctx,
-    uint64_t seq = ctx->rqt->get_job_counter ();
     int rc = -1;
 
+    if (!ctx || !ctx->rqt) {
+        errno = EINVAL;
+        return -1;
+    }
```
Good catch!

trws commented on resource/reapi/bindings/c/reapi_cli.cpp:
```diff
@@ -393,13 +417,15 @@ extern "C" int reapi_cli_stat (reapi_cli_ctx_t *ctx,
+        errno = EINVAL;
+        return strdup ("ERROR: reapi_cli context is null\n");
+    }
 
+    std::string err_buf;
     if (ctx->rqt)
-        err_buf = ctx->rqt->get_resource_query_err_msg () + reapi_cli_t::get_err_message ()
-                  + ctx->err_msg;
-    else
-        err_buf = reapi_cli_t::get_err_message () + ctx->err_msg;
+        err_buf = ctx->rqt->get_resource_query_err_msg ();
+    err_buf += reapi_cli_t::get_err_message () + ctx->err_msg;
 
     return strdup (err_buf.c_str ());
```
This is right given the contract we had, but I'm wondering if we should look at making this not return a string the caller has to free.

milroy replied:
Yeah, that makes sense. What about adding a commit that does something like this:

```C++
struct reapi_cli_ctx {
    resource_query_t *rqt;
    std::string err_msg;
    std::string err_out;   // return buffer for reapi_cli_get_err_msg
};

extern "C" const char *reapi_cli_get_err_msg (reapi_cli_ctx_t *ctx)
{
    if (!ctx) {
        errno = EINVAL;
        return "ERROR: reapi_cli context is null\n";   // static storage; not freed
    }
    ctx->err_out.clear ();
    if (ctx->rqt)
        ctx->err_out = ctx->rqt->get_resource_query_err_msg ();
    ctx->err_out += reapi_cli_t::get_err_message () + ctx->err_msg;
    return ctx->err_out.c_str ();
}
```

trws replied:
How about we make that a new PR? I think this one is ready to go as is.

trws reviewed (approved):
Looks great, thanks for the quick rework @milroy!
